src/onnx/parse_einsum.cpp

```diff
@@ -0,0 +1,774 @@
+        op = unsqueeze_transpose(info, cur_pair, op);
+
+        std::vector<int> red;
+        // Check if a given label appears in any of the subsequent mapping matrix terms(this
+        // includes the output). If does not, it is reduced and marked as -1 in cur_pair.
+        for(int d = 0; d < map_mat[0].size(); ++d)
+        {
+            bool all_neg_one = all_of(extract_column(map_mat, d, input_idx + 1, map_mat.size()),
+                                      [](auto i) { return i == -1; });
+            if(all_neg_one and cur_pair[1][d] != -1 and cur_pair[0][d] == -1)
+                red.push_back(d);
+        }
+
+        return apply_reduce_sum_op(info, op, red, cur_pair[1]);
+    }
+
+    instruction_ref gather_diagonal(const onnx_parser::node_info& info,
+                                    int_mat& cur_pair,
+                                    instruction_ref op,
+                                    const int_mat& diag) const
+    {
+        if(diag.size() != 1)
+            MIGRAPHX_THROW(
+                "Parsing of equations with more than one duplicated labels per input term is not "
+                "implemented");
+
+        const auto& op_lens = op->get_shape().lens();
+
+        int first_axis               = diag[0][0];
+        const std::vector<int>& axes = diag[0];
+        if(not all_of(axes, [&](int a) { return op_lens[first_axis] == op_lens[a]; }))
+            MIGRAPHX_THROW("All duplicate labels have to be the same dimension");
+
+        std::vector<int> batch_axes = set_difference(arange(0, op_lens.size()), axes);
+        if(not all_of(batch_axes, [&](int ba) { return ba < axes.front(); }))
+            MIGRAPHX_THROW(
+                "Parsing of equations with duplicated labels and batch axes that are not "
+                "the outer-most axes, is not implemented");
+
+        size_t batch_size = calc_dim(batch_axes, op_lens);
+
+        std::vector<size_t> indices;
+        for(size_t batch = 0; batch < batch_size; ++batch)
+        {
+            for(size_t i = 0; i < op_lens[first_axis]; ++i)
+            {
+                std::vector<size_t> index(axes.size(), i);
+                indices.insert(indices.end(), index.begin(), index.end());
+            }
+        }
+
+        std::vector<size_t> indices_lens{op_lens[first_axis], axes.size()};
+        if(batch_size > 1)
+            indices_lens.insert(indices_lens.begin(), batch_size);
+
+        auto indices_arg = info.add_literal(
+            migraphx::literal{migraphx::shape{migraphx::shape::int64_type, indices_lens}, indices});
+
+        op = info.add_instruction(
+            migraphx::make_op("gathernd", {{"batch_dims", batch_axes.size()}}), op, indices_arg);
+
+        // compute output row
+        for(auto& r : cur_pair[1])
+            if(contains(axes, r))
+                r = first_axis;
+
+        std::vector<int> to_remove(axes.begin() + 1, axes.end());
+        for(auto t : to_remove)
```

Comment:
Just do `range(axes.begin() + 1, axes.end())` instead of constructing the intermediate vector.